Description: Hide the first level of directories
 in the recordings menu, all recordings are stored at
 the "local" directory
Origin: www.vdr-portal.de/board17-developer/board97-vdr-core/p1177898-/#post1177898
Forwarded: no
Author: Lars Hanisch <dvb@flensrocker.de>

--- a/recording.c
+++ b/recording.c
@@ -44,9 +44,9 @@
 #define NAMEFORMAT   "%s/%s/" DATAFORMAT
 */
 #define DATAFORMATPES   "%4d-%02d-%02d.%02d%*c%02d.%02d.%02d" RECEXT
-#define NAMEFORMATPES   "%s/%s/" "%4d-%02d-%02d.%02d.%02d.%02d.%02d" RECEXT
+#define NAMEFORMATPES   "%s/%s%s/" "%4d-%02d-%02d.%02d.%02d.%02d.%02d" RECEXT
 #define DATAFORMATTS    "%4d-%02d-%02d.%02d.%02d.%d-%d" RECEXT
-#define NAMEFORMATTS    "%s/%s/" DATAFORMATTS
+#define NAMEFORMATTS    "%s/%s%s/" DATAFORMATTS
 
 #define RESUMEFILESUFFIX  "/resume%s%s"
 #ifdef SUMMARYFALLBACK
@@ -758,6 +758,9 @@
   sortBufferName = sortBufferTime = NULL;
   fileName = NULL;
   name = NULL;
+  firstLevelFolderIfHidden = "";
+  if (cVideoDirectory::HideFirstRecordingLevel())
+     firstLevelFolderIfHidden = "local/";
   fileSizeMB = -1; // unknown
   channel = Timer->Channel()->Number();
   instanceId = InstanceId;
@@ -829,6 +832,7 @@
   if (strstr(FileName, cVideoDirectory::Name()) == FileName)
      FileName += strlen(cVideoDirectory::Name()) + 1;
   const char *p = strrchr(FileName, '/');
+  firstLevelFolderIfHidden = "";
 
   name = NULL;
   info = new cRecordingInfo(fileName);
@@ -843,9 +847,18 @@
         t.tm_mon--;
         t.tm_sec = 0;
         start = mktime(&t);
-        name = MALLOC(char, p - FileName + 1);
-        strncpy(name, FileName, p - FileName);
-        name[p - FileName] = 0;
+        const char *copyFileName = FileName;
+        if (cVideoDirectory::HideFirstRecordingLevel()) {
+           const char *f = strchr(FileName, '/');
+           if ((f != NULL) && (f < p)) {
+              copyFileName = f + 1;
+              firstLevelFolderIfHidden = FileName;
+              firstLevelFolderIfHidden.Truncate(f - FileName + 1);
+              }
+           }
+        name = MALLOC(char, p - copyFileName + 1);
+        strncpy(name, copyFileName, p - copyFileName);
+        name[p - copyFileName] = 0;
         name = ExchangeChars(name, false);
         isPesRecording = instanceId < 0;
         }
@@ -982,7 +995,7 @@
         *sb = strdup(buf);
         }
      else {
-        char *s = strdup(FileName() + strlen(cVideoDirectory::Name()));
+        char *s = strdup(FileName() + strlen(cVideoDirectory::Name()) + strlen(*firstLevelFolderIfHidden));
         if (RecordingsSortMode != rsmName || Setup.AlwaysSortFoldersFirst)
            s = StripEpisodeName(s, RecordingsSortMode != rsmName);
         strreplace(s, '/', '0'); // some locales ignore '/' when sorting
@@ -1051,7 +1064,7 @@
      if (strcmp(Name, name) != 0)
         dsyslog("recording file name '%s' truncated to '%s'", name, Name);
      Name = ExchangeChars(Name, true);
-     fileName = strdup(cString::sprintf(fmt, cVideoDirectory::Name(), Name, t->tm_year + 1900, t->tm_mon + 1, t->tm_mday, t->tm_hour, t->tm_min, ch, ri));
+     fileName = strdup(cString::sprintf(fmt, cVideoDirectory::Name(), *firstLevelFolderIfHidden, Name, t->tm_year + 1900, t->tm_mon + 1, t->tm_mday, t->tm_hour, t->tm_min, ch, ri));
      free(Name);
      }
   return fileName;
--- a/recording.h
+++ b/recording.h
@@ -105,6 +105,7 @@
   mutable char *sortBufferTime;
   mutable char *fileName;
   mutable char *name;
+  cString firstLevelFolderIfHidden;
   mutable int fileSizeMB;
   mutable int numFrames;
   int channel;
--- a/vdr.c
+++ b/vdr.c
@@ -262,6 +262,7 @@
       { "genindex", required_argument, NULL, 'g' | 0x100 },
       { "grab",     required_argument, NULL, 'g' },
       { "help",     no_argument,       NULL, 'h' },
+      { "hide-first-recording-level", no_argument, NULL, 'H' },
       { "instance", required_argument, NULL, 'i' },
       { "lib",      required_argument, NULL, 'L' },
       { "lirc",     optional_argument, NULL, 'l' | 0x100 },
@@ -288,7 +289,7 @@
     };
 
   int c;
-  while ((c = getopt_long(argc, argv, "a:c:dD:e:E:g:hi:l:L:mp:P:r:s:t:u:v:Vw:", long_options, NULL)) != -1) {
+  while ((c = getopt_long(argc, argv, "a:c:dD:e:E:g:hHi:l:L:mp:P:r:s:t:u:v:Vw:", long_options, NULL)) != -1) {
         switch (c) {
           case 'a': AudioCommand = optarg;
                     break;
@@ -376,6 +377,8 @@
                     break;
           case 'h': DisplayHelp = true;
                     break;
+          case 'H': cVideoDirectory::SetHideFirstRecordingLevel(true);
+                    break;
           case 'i': if (isnumber(optarg)) {
                        InstanceId = atoi(optarg);
                        if (InstanceId >= 0)
@@ -561,6 +564,13 @@
                "                           existing directory, without any \"..\", double '/'\n"
                "                           or symlinks (default: none, same as -g-)\n"
                "  -h,       --help         print this help and exit\n"
+               "  -H,       --hide-first-recording-level\n"
+               "                           The first level directories in the video directory\n"
+               "                           will be ignored in the recordings menu.\n"
+               "                           All recordings will be placed in the directory 'local'.\n"
+               "                           You can mount/link other video directories inside the\n"
+               "                           video directory to display their contents side by side\n"
+               "                           with your local video directory\n"
                "  -i ID,    --instance=ID  use ID as the id of this VDR instance (default: 0)\n"
                "  -l LEVEL, --log=LEVEL    set log level (default: 3)\n"
                "                           0 = no logging, 1 = errors only,\n"
--- a/videodir.c
+++ b/videodir.c
@@ -20,6 +20,7 @@
 #include "tools.h"
 
 cString cVideoDirectory::name;
+bool cVideoDirectory::hide_first_recording_level = false;
 cVideoDirectory *cVideoDirectory::current = NULL;
 
 cVideoDirectory::cVideoDirectory(void)
--- a/videodir.h
+++ b/videodir.h
@@ -16,6 +16,7 @@
 class cVideoDirectory {
 private:
   static cString name;
+  static bool hide_first_recording_level;
   static cVideoDirectory *current;
   static cVideoDirectory *Current(void);
 public:
@@ -81,6 +82,8 @@
   static cString PrefixVideoFileName(const char *FileName, char Prefix);
   static void RemoveEmptyVideoDirectories(const char *IgnoreFiles[] = NULL);
   static bool IsOnVideoDirectoryFileSystem(const char *FileName);
+  static void SetHideFirstRecordingLevel(bool Hide) { hide_first_recording_level = Hide; };
+  static bool HideFirstRecordingLevel(void) { return hide_first_recording_level; };
   };
 
 class cVideoDiskUsage {
